packages/dd-trace/src/encode/0.5.js

```diff
@@ -1,101 +1,235 @@
+  _encode (bytes, trace) {
+    this._encodeArrayPrefix(bytes, trace)
+
+    for (const span of trace) {
+      this._encodeByte(bytes, ARRAY_OF_TWELVE)
+      this._encodeString(bytes, span.service)
+      this._encodeString(bytes, span.name)
+      this._encodeString(bytes, span.resource)
+      this._encodeId(bytes, span.trace_id)
+      this._encodeId(bytes, span.span_id)
+      this._encodeId(bytes, span.parent_id)
+      this._encodeInteger(bytes, span.start || 0)
+      this._encodeInteger(bytes, span.duration || 0)
+      this._encodeInteger(bytes, span.error)
+      this._encodeMap(bytes, span.meta || {})
+      this._encodeMap(bytes, span.metrics || {})
+      this._encodeString(bytes, span.type)
+    }
+  }
+
+  _reset () {
+    this._traces = []
+    this._traceSize = 0
+    this._stringDefers = [{ value: '', length: 0, prefix: [0xa0] }]
+    this._stringMap = { '': 0 }
+    this._stringSize = 1
+  }
+
+  _encodeArrayPrefix (bytes, value) {
+    const length = value.length
+
+    if (length < 0x10) { // fixarray
+      bytes.push(length | 0x90)
+    } else if (length < 0x10000) { // array 16
+      bytes.push(0xdc, length >> 8, length)
+    } else if (length < 0x100000000) { // array 32
+      bytes.push(0xdd, length >> 24, length >> 16, length >> 8, length)
+    } else {
+      throw new Error('Array too large')
+    }
+  }
+
+  _encodeByte (bytes, value) {
+    bytes.push(value)
+  }
+
+  _encodeId (bytes, id) {
+    id = id.toArray()
+    bytes.push(0xcf, id[0], id[1], id[2], id[3], id[4], id[5], id[6], id[7])
+  }
+
+  _encodeInteger (bytes, value) {
+    notepack._encode(bytes, [], value)
+  }
 
-function init () {
-  strings = Buffer.allocUnsafe(EncoderState.MAX_SIZE)
-  stringMap = {}
-  stringMapLen = 0
-  stringsBufLen = 3 // 0xdc and then uint16
-  strings[0] = 0xdc
+  _encodeMap (bytes, value) {
+    const keys = []
+    const allKeys = Object.keys(value)
+
+    let key = ''
+
+    for (let i = 0, l = allKeys.length; i < l; i++) {
+      key = allKeys[i]
+      if (typeof value[key] !== 'function') {
+        keys.push(key)
+      }
+    }
+
+    const length = keys.length
+
+    if (length < 0x10) { // fixmap
+      bytes.push(length | 0x80)
+    } else if (length < 0x10000) { // map 16
+      bytes.push(0xde, length >> 8, length)
+    } else if (length < 0x100000000) { // map 32
+      bytes.push(0xdf, length >> 24, length >> 16, length >> 8, length)
+    } else {
+      throw new Error('Object too large')
+    }
+
+    for (let i = 0; i < length; i++) {
+      key = keys[i]
+      this._encodeString(bytes, key)
+      this._encodeValue(bytes, value[key])
+    }
+  }
+
+  _encodeValue (bytes, value) {
+    switch (typeof value) {
+      case 'string':
+        this._encodeString(bytes, value)
+        break
+      case 'number':
+        this._encodeFloat(bytes, value)
+        break
+      default:
+        // should not happen
+    }
+  }
+
+  _encodeString (bytes, value = '') {
+    if (!(value in this._stringMap)) {
```

Comment:
In general, it's my understanding that it's faster to retrieve the value, and eventually return that retrieved value, than to check with `in` and then retrieve it again on line 72.

Like this:

```js
let stringKey = this._stringMap[value]
if (typeof stringKey !== 'string') {
  // ....
  stringKey = / * ... */
  this._stringMap[value] = stringKey
}
this._encodeInteger(bytes, stringKey)
```
}

Reply:
I think the compiler is able to optimize accessing the same property multiple times within a function. Also, for some reason switching to your approach results a significant negative impact (~70% slower encoding).

Reply:
> switching to your approach results a significant negative impact (~70% slower encoding)

K let's not do that then 😄 .